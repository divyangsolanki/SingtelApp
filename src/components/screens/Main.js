import React, {useState, useEffect} from 'react';
import {StyleSheet, Text, View, ScrollView, Alert} from 'react-native';
import Card from '../common/Card';
import ButtonClickable from '../common/ButtonClickable';

const Main = props => {
  const CARD_PAIRS_VALUE = props.CARD_PAIRS_VALUE;
  const TOTAL_CARDS = props.TOTAL_CARDS;

  const duplicateCard = () => {
    return CARD_PAIRS_VALUE.reduce((preValue, current, index, array) => {
      return preValue.concat([current, current]);
    }, []);
  };

  const [isFlipped, setIsFlipped] = useState(Array(TOTAL_CARDS).fill(false));
  const shuffle = duplicateCard().sort(() => Math.random() - 0.5);
  const [shuffledCard, setShuffleCard] = useState(shuffle);
  const [clickCount, setClickCount] = useState(1);
  const [prevSelectedCard, setPrevSelectedCard] = useState(-1);
  const [prevCardId, setPrevCardId] = useState(-1);
  const [moves, setMoves] = useState(0);

  const handleClick = (cardNum, index) => {
    const cardId = index;
    const newFlipps = isFlipped.slice();
    setPrevSelectedCard(shuffledCard[cardId]);
    setPrevCardId(cardId);
    if (newFlipps[cardId] === false) {
      newFlipps[cardId] = true;
      setIsFlipped(prevState => newFlipps);
      setClickCount(clickCount + 1);
      if (clickCount === 2) {
        setClickCount(1);
        setMoves(moves + 1);
        const newCard = shuffledCard[cardId];
        const previousCard = prevSelectedCard;
        setTimeout(() => {
          isCardMatch(previousCard, newCard, prevCardId, cardId, newFlipps);
        }, 500);
      }
    }
  };
  useEffect(() => {
    if (isGameOver()) {
      Alert.alert(
        'Congratulations..!!',
        'You have finished the game with ' + moves + ' moves.',
        [
          {
            text: 'OK',
            onPress: () => {
              restartGame();
            },
          },
        ],
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isFlipped]);

  const isCardMatch = (card1, card2, card1Id, card2Id, newFlipps) => {
    if (card1 === card2) {
      const hideCard = shuffledCard.slice();
      hideCard[card1Id] = card1;
      hideCard[card2Id] = card2;
      setIsFlipped(prevState => newFlipps);
      setTimeout(() => {
        setShuffleCard(hideCard);
      }, 500);
    } else {
      const flipBack = isFlipped.slice();
      flipBack[card1Id] = false;
      flipBack[card2Id] = false;
      setTimeout(() => {
        setIsFlipped(flipBack);
      }, 500);
    }
  };

  const restartGame = () => {
    setIsFlipped(Array(TOTAL_CARDS).fill(false));
    setShuffleCard(shuffle);
    setClickCount(1);
    setPrevSelectedCard(-1);
    setPrevCardId(-1);
    setMoves(0);
  };

  const isGameOver = () => {
    return isFlipped.every((element, index, array) => element !== false);
  };

  return (
    <View style={styles.container}>
      <Text style={styles.flipText}>Flip Card</Text>
      <ScrollView style={styles.scrollViewContainer}>
        <View style={styles.cardContainer}>
          {shuffledCard.map((cardNum, index) => (
            <Card
              key={index}
              val={cardNum}
              testID="card"
              onPress={() => handleClick(cardNum, index)}
              isFlipped={isFlipped[index]}
            />
          ))}
        </View>
      </ScrollView>
      <View style={styles.bottomContainer}>
        <Text style={styles.movesText}>Moves: {moves}</Text>
        <View style={styles.restartButtonContainer}>
          <ButtonClickable testID="restart" onPress={() => restartGame()}>
            Restart
          </ButtonClickable>
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#F5FCFF',
    marginTop: 20,
  },
  scrollViewContainer: {
    flex: 0.85,
  },
  cardContainer: {
    margin: 5,
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
  bottomContainer: {
    flex: 0.15,
  },
  restartButtonContainer: {
    alignSelf: 'center',
  },
  movesText: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  flipText: {
    fontSize: 20,
    textAlign: 'center',
  },
  instructions: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5,
  },
  card: {
    width: 200,
  },
  face: {
    flex: 1,
    backgroundColor: '#2ecc71',
    justifyContent: 'center',
    alignItems: 'center',
  },
  back: {
    flex: 1,
    backgroundColor: '#f1c40f',
    justifyContent: 'center',
    alignItems: 'center',
  },
  button: {
    width: 100,
    height: 30,
    marginTop: 30,
    paddingTop: 6,
    paddingBottom: 6,
    borderRadius: 3,
    borderWidth: 1,
    backgroundColor: '#007AFF',
    borderColor: 'transparent',
  },
  buttonText: {
    color: '#fff',
    textAlign: 'center',
  },
  img: {
    flex: 1,
    height: 64,
  },
});

export default Main;
