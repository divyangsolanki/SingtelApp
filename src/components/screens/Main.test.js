import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import Main from './Main';

describe('Test Main Component', () => {
  it('Main snapshot', () => {
    const wrapper = render(<Main CARD_PAIRS_VALUE={[0, 1]} TOTAL_CARDS={4} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  it('Main handleClick [Matched card]', () => {
    const wrapper = render(<Main CARD_PAIRS_VALUE={[0]} TOTAL_CARDS={2} />);
    const {getAllByTestId} = wrapper;
    fireEvent.press(getAllByTestId('card')[0]);
    fireEvent.press(getAllByTestId('card')[1]);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  it('Main handleClick [Unmatched card]', () => {
    const wrapper = render(<Main CARD_PAIRS_VALUE={[0, 1]} TOTAL_CARDS={4} />);
    const {getAllByTestId} = wrapper;
    fireEvent.press(getAllByTestId('card')[0]);
    fireEvent.press(getAllByTestId('card')[1]);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  it('Main restart', () => {
    const wrapper = render(<Main CARD_PAIRS_VALUE={[0]} TOTAL_CARDS={2} />);
    const {getByTestId} = wrapper;
    fireEvent.press(getByTestId('restart'));
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
