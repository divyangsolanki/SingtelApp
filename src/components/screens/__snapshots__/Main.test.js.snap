// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Main Component Main handleClick [Matched card] 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "marginTop": 20,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
        "textAlign": "center",
      }
    }
  >
    Flip Card
  </Text>
  <RCTScrollView
    style={
      Object {
        "flex": 0.85,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "margin": 5,
          }
        }
      >
        <View
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 2,
              "flexWrap": "wrap",
              "margin": 10,
            }
          }
          testID="card"
        />
        <View
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 2,
              "flexWrap": "wrap",
              "margin": 10,
            }
          }
          testID="card"
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "flex": 0.15,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
          "margin": 10,
          "textAlign": "center",
        }
      }
    >
      Moves: 
      1
    </Text>
    <View
      style={
        Object {
          "alignSelf": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#003566",
            "borderRadius": 32,
            "height": 64,
            "justifyContent": "center",
            "width": 260,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "alignSelf": "center",
                "color": "#ffffff",
                "fontSize": 14,
                "fontStyle": "normal",
                "letterSpacing": 0,
                "textAlign": "center",
              },
              undefined,
              Object {
                "position": "absolute",
                "width": 260,
              },
            ]
          }
        >
          Restart
        </Text>
        <View
          accessibilityRole="button"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="restart"
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
                                              
                                
                                
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Test Main Component Main handleClick [Unmatched card] 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "marginTop": 20,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
        "textAlign": "center",
      }
    }
  >
    Flip Card
  </Text>
  <RCTScrollView
    style={
      Object {
        "flex": 0.85,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "margin": 5,
          }
        }
      >
        <View
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 2,
              "flexWrap": "wrap",
              "margin": 10,
            }
          }
          testID="card"
        />
        <View
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 2,
              "flexWrap": "wrap",
              "margin": 10,
            }
          }
          testID="card"
        />
        <View
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 2,
              "flexWrap": "wrap",
              "margin": 10,
            }
          }
          testID="card"
        />
        <View
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 2,
              "flexWrap": "wrap",
              "margin": 10,
            }
          }
          testID="card"
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "flex": 0.15,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
          "margin": 10,
          "textAlign": "center",
        }
      }
    >
      Moves: 
      1
    </Text>
    <View
      style={
        Object {
          "alignSelf": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#003566",
            "borderRadius": 32,
            "height": 64,
            "justifyContent": "center",
            "width": 260,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "alignSelf": "center",
                "color": "#ffffff",
                "fontSize": 14,
                "fontStyle": "normal",
                "letterSpacing": 0,
                "textAlign": "center",
              },
              undefined,
              Object {
                "position": "absolute",
                "width": 260,
              },
            ]
          }
        >
          Restart
        </Text>
        <View
          accessibilityRole="button"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="restart"
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
                                              
                                
                                
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Test Main Component Main restart 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "marginTop": 20,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
        "textAlign": "center",
      }
    }
  >
    Flip Card
  </Text>
  <RCTScrollView
    style={
      Object {
        "flex": 0.85,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "margin": 5,
          }
        }
      >
        <View
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 2,
              "flexWrap": "wrap",
              "margin": 10,
            }
          }
          testID="card"
        />
        <View
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 2,
              "flexWrap": "wrap",
              "margin": 10,
            }
          }
          testID="card"
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "flex": 0.15,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
          "margin": 10,
          "textAlign": "center",
        }
      }
    >
      Moves: 
      0
    </Text>
    <View
      style={
        Object {
          "alignSelf": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#003566",
            "borderRadius": 32,
            "height": 64,
            "justifyContent": "center",
            "width": 260,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "alignSelf": "center",
                "color": "#ffffff",
                "fontSize": 14,
                "fontStyle": "normal",
                "letterSpacing": 0,
                "textAlign": "center",
              },
              undefined,
              Object {
                "position": "absolute",
                "width": 260,
              },
            ]
          }
        >
          Restart
        </Text>
        <View
          accessibilityRole="button"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="restart"
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
                                              
                                
                                
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Test Main Component Main snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "marginTop": 20,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
        "textAlign": "center",
      }
    }
  >
    Flip Card
  </Text>
  <RCTScrollView
    style={
      Object {
        "flex": 0.85,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "margin": 5,
          }
        }
      >
        <View
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 2,
              "flexWrap": "wrap",
              "margin": 10,
            }
          }
          testID="card"
        />
        <View
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 2,
              "flexWrap": "wrap",
              "margin": 10,
            }
          }
          testID="card"
        />
        <View
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 2,
              "flexWrap": "wrap",
              "margin": 10,
            }
          }
          testID="card"
        />
        <View
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 2,
              "flexWrap": "wrap",
              "margin": 10,
            }
          }
          testID="card"
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "flex": 0.15,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
          "margin": 10,
          "textAlign": "center",
        }
      }
    >
      Moves: 
      0
    </Text>
    <View
      style={
        Object {
          "alignSelf": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#003566",
            "borderRadius": 32,
            "height": 64,
            "justifyContent": "center",
            "width": 260,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "alignSelf": "center",
                "color": "#ffffff",
                "fontSize": 14,
                "fontStyle": "normal",
                "letterSpacing": 0,
                "textAlign": "center",
              },
              undefined,
              Object {
                "position": "absolute",
                "width": 260,
              },
            ]
          }
        >
          Restart
        </Text>
        <View
          accessibilityRole="button"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="restart"
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
                                              
                                
                                
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
